File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/logic/localization_logic.cpp
#include "crtp_cpp/logic/localization_logic.hpp"
#include <algorithm>
#include <stdexcept>

LocalizationLogic::LocalizationLogic(CrtpLink* crtp_link)
    : Logic(crtp_link),
      packer(LocalizationPacker()) {}

void LocalizationLogic::send_extpos(const std::vector<float>& pos) {
    CrtpRequest request;
    request.packet = packer.send_extpos(pos[0], pos[1], pos[2]);
    link->send_packet_no_response(request);
}

void LocalizationLogic::send_short_lpp_packet(uint8_t dest_id, const std::vector<uint8_t>& data) {
    CrtpRequest request;
    request.packet = packer.send_short_lpp_packet(dest_id, data);
    link->send_packet_no_response(request);
}

void LocalizationLogic::send_emergency_stop() {
    CrtpRequest request;
    request.packet = packer.send_emergency_stop();
    link->send_packet_no_response(request);
}

void LocalizationLogic::send_emergency_stop_watchdog() {
    CrtpRequest request;
    request.packet = packer.send_emergency_stop_watchdog();
    link->send_packet_no_response(request);
}

void LocalizationLogic::send_extpose(const std::vector<float>& pos, const std::vector<float>& quat) {
    CrtpRequest request;
    request.packet = packer.send_extpose(pos[0], pos[1], pos[2], quat[0], quat[1], quat[2], quat[3]);
    link->send_packet_no_response(request);
}

void LocalizationLogic::send_lh_persist_data_packet(const std::vector<int>& geo_list, const std::vector<int>& calib_list) {
    std::vector<int> sorted_geo_list = geo_list;
    std::vector<int> sorted_calib_list = calib_list;
    std::sort(sorted_geo_list.begin(), sorted_geo_list.end());
    std::sort(sorted_calib_list.begin(), sorted_calib_list.end());

    int max_bs_nr = 15;
    if (!sorted_geo_list.empty()) {
        if (sorted_geo_list.front() < 0 || sorted_geo_list.back() > max_bs_nr) {
            throw std::runtime_error("Geometry BS list is not valid");
        }
    }
    if (!sorted_calib_list.empty()) {
        if (sorted_calib_list.front() < 0 || sorted_calib_list.back() > max_bs_nr) {
            throw std::runtime_error("Calibration BS list is not valid");
        }
    }

    uint16_t mask_geo = 0;
    uint16_t mask_calib = 0;
    for (int bs : sorted_geo_list) {
        mask_geo += 1 << bs;
    }
    for (int bs : sorted_calib_list) {
        mask_calib += 1 << bs;
    }

    CrtpRequest request;
    request.packet = packer.send_lh_persist_data_packet(mask_geo, mask_calib);
    link->send_packet_no_response(request);
}

std::map<std::string, std::vector<float>> LocalizationLogic::_decode_lh_angle(const std::vector<uint8_t>& data) {
    std::map<std::string, std::vector<float>> decoded_data;

    if (data.size() != 26) {
        throw std::runtime_error("Invalid data size for LH angle decoding");
    }

    float raw_data[9];
    std::memcpy(raw_data, data.data(), 26);

    decoded_data["basestation"] = {raw_data[0]};
    decoded_data["x"] = {raw_data[1], raw_data[1] - fp16_to_float(raw_data[2]), raw_data[1] - fp16_to_float(raw_data[3]), raw_data[1] - fp16_to_float(raw_data[4])};
    decoded_data["y"] = {raw_data[5], raw_data[5] - fp16_to_float(raw_data[6]), raw_data[5] - fp16_to_float(raw_data[7]), raw_data[5] - fp16_to_float(raw_data[8])};

    return decoded_data;
}

float LocalizationLogic::fp16_to_float(uint16_t fp16){
    uint32_t s = (fp16 >> 15) & 0x00000001; // sign
    uint32_t e = (fp16 >> 10) & 0x0000001f; // exponent
    uint32_t f = fp16 & 0x000003ff;         // fraction

    if (e == 0) {
        if (f == 0) {
            uint32_t result = s << 31;
            float float_result;
            std::memcpy(&float_result, &result, sizeof(float));
            return float_result;
        } else {
            while (!(f & 0x00000400)) {
                f <<= 1;
                e -= 1;
            }
            e += 1;
            f &= ~0x00000400;
        }
    } else if (e == 31) {
        if (f == 0) {
            uint32_t result = (s << 31) | 0x7f800000;
            float float_result;
            std::memcpy(&float_result, &result, sizeof(float));
            return float_result;
        } else {
            uint32_t result = (s << 31) | 0x7f800000 | (f << 13);
            float float_result;
            std::memcpy(&float_result, &result, sizeof(float));
            return float_result;
        }
    }

    e += 127 - 15;
    f <<= 13;
    uint32_t result = (s << 31) | (e << 23) | f;
    float float_result;
    std::memcpy(&float_result, &result, sizeof(float));
    return float_result;
}

File: src/dependencies/crazyflie_hardware/src/crtp/crtp/logic/localization_logic.py
import struct
from .logic import Logic
from crtp.crtp_link import CrtpLink
from crtp.packers.localization_packer import LocalizationPacker
from crtp.packers.crtp_packer import CrtpPacker

from crtp.utils.encoding import fp16_to_float

from typing import Callable


class LocalizationLogic(Logic):
    """
    Handle localization-related data communication with the Crazyflie
    """

    def __init__(
        self, crtp_packer_factory: Callable[[int], CrtpPacker], crtp_link: CrtpLink
    ):
        super().__init__(crtp_link)
        self.packer = LocalizationPacker(crtp_packer_factory)

    def send_extpos(self, pos):
        """
        Send the current Crazyflie X, Y, Z position. This is going to be
        forwarded to the Crazyflie's position estimator.
        """
        packet = self.packer.send_extpos(pos)
        self.link.send_packet_no_response(packet)

    def send_short_lpp_packet(self, dest_id, data):
        """
        Send ultra-wide-band LPP packet to dest_id
        """
        packet = self.packer.send_short_lpp_packet(dest_id, data)
        self.link.send_packet_no_response(packet)

    def send_emergency_stop(self):
        """
        Send emergency stop
        """
        packet = self.packer.send_emergency_stop()
        self.link.send_packet_no_response(packet)

    def send_emergency_stop_watchdog(self):
        """
        Send emergency stop watchdog
        """
        packet = self.packer.send_emergency_stop_watchdog()
        self.link.send_packet_no_response(packet)

    def send_extpose(self, pos, quat):
        """
        Send the current Crazyflie pose (position [x, y, z] and
        attitude quaternion [qx, qy, qz, qw]). This is going to be forwarded
        to the Crazyflie's position estimator.
        """
        packet = self.packer.send_extpose(pos, quat)
        self.link.send_packet_no_response(packet)

    def send_lh_persist_data_packet(self, geo_list, calib_list):
        """
        Send geometry and calibration data to persistent memory subsystem
        """
        geo_list.sort()
        calib_list.sort()
        max_bs_nr = 15
        if len(geo_list) > 0:
            if geo_list[0] < 0 or geo_list[-1] > max_bs_nr:
                raise Exception("Geometry BS list is not valid")
        if len(calib_list) > 0:
            if calib_list[0] < 0 or calib_list[-1] > max_bs_nr:
                raise Exception("Calibration BS list is not valid")

        mask_geo = 0
        mask_calib = 0
        for bs in geo_list:
            mask_geo += 1 << bs
        for bs in calib_list:
            mask_calib += 1 << bs

        packet = self.packer.send_lh_persist_data_packet(mask_geo, mask_calib)
        self.link.send_packet_no_response(packet)

    def _decode_lh_angle(self, data):
        decoded_data = {}

        raw_data = struct.unpack("<Bfhhhfhhh", data)

        decoded_data["basestation"] = raw_data[0]
        decoded_data["x"] = [0, 0, 0, 0]
        decoded_data["x"][0] = raw_data[1]
        decoded_data["x"][1] = raw_data[1] - fp16_to_float(raw_data[2])
        decoded_data["x"][2] = raw_data[1] - fp16_to_float(raw_data[3])
        decoded_data["x"][3] = raw_data[1] - fp16_to_float(raw_data[4])
        decoded_data["y"] = [0, 0, 0, 0]
        decoded_data["y"][0] = raw_data[5]
        decoded_data["y"][1] = raw_data[5] - fp16_to_float(raw_data[6])
        decoded_data["y"][2] = raw_data[5] - fp16_to_float(raw_data[7])
        decoded_data["y"][3] = raw_data[5] - fp16_to_float(raw_data[8])

        return decoded_data
